{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Problem 1: Controls\n",
    "\n",
    "This Python script creates two lists with the first values of each row for both files. Then takes them in pairs and compares them to see if they are in the same sequence or not. "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 39,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "All under control!\n"
     ]
    }
   ],
   "source": [
    "import csv\n",
    "\n",
    "agi_lg_list = []\n",
    "agi_g_list = []\n",
    "\n",
    "with open(\"/home/osboxes/BioinformaticsCourseGit/LOM_BioinfoIntro/EXAMS_W2/LocusGene.tsv\") as lg_tsvfile:\n",
    "    lg = csv.DictReader(lg_tsvfile, fieldnames = (\"GermplasmLocus\",\"Gene\",\"ProteinLength\"), delimiter = \"\\t\")\n",
    "    for row in lg:\n",
    "        agi_lg_list.append(row[\"GermplasmLocus\"])\n",
    "\n",
    "with open(\"/home/osboxes/BioinformaticsCourseGit/LOM_BioinfoIntro/EXAMS_W2/Germplasm.tsv\") as g_tsvfile:\n",
    "    g = csv.DictReader(g_tsvfile, fieldnames= (\"Locus\",\"Germplasm\",\"Phenotype\",\"Pubmed\"), delimiter = \"\\t\")\n",
    "    for row in g:\n",
    "        agi_g_list.append(row[\"Locus\"])\n",
    "        \n",
    "for index in range(len(agi_lg_list)):\n",
    "    if agi_lg_list[index] == agi_g_list[index]:\n",
    "        pass\n",
    "    else:\n",
    "        print(agi_lg_list[index] + \" and \" + agi_g_list[index])\n",
    "        \n",
    "if int(index) == int(len(agi_lg_list)-1):\n",
    "    print(\"All under control!\")"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Problem 2: Design and create database\n",
    "\n",
    "This is a database called \"EW2\" (for Exam Week 2) which contains two tables, Germplasm and Locus_Gene.\n",
    "Germplasm has 4 variables apart from the ID number (Locus, Germplasm, Phenotype and Pubmed) and Locus_Gene has 3 variables apart from the ID number (Germplasm_Locus, Gene, Protein_Length)."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "'Connected: root@mysql'"
      ]
     },
     "execution_count": 4,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "%load_ext sql\n",
    "%config SqlMagic.autocommit=False\n",
    "%sql mysql+pymysql://root:root@127.0.0.1:3306/mysql"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 91,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      " * mysql+pymysql://root:***@127.0.0.1:3306/mysql\n",
      "6 rows affected.\n",
      " * mysql+pymysql://root:***@127.0.0.1:3306/mysql\n",
      "0 rows affected.\n",
      " * mysql+pymysql://root:***@127.0.0.1:3306/mysql\n",
      "0 rows affected.\n",
      " * mysql+pymysql://root:***@127.0.0.1:3306/mysql\n",
      "0 rows affected.\n",
      " * mysql+pymysql://root:***@127.0.0.1:3306/mysql\n",
      "0 rows affected.\n",
      " * mysql+pymysql://root:***@127.0.0.1:3306/mysql\n",
      "0 rows affected.\n",
      " * mysql+pymysql://root:***@127.0.0.1:3306/mysql\n",
      "2 rows affected.\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<table>\n",
       "    <tr>\n",
       "        <th>Tables_in_EW2</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "        <td>Germplasm</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "        <td>Locus_Gene</td>\n",
       "    </tr>\n",
       "</table>"
      ],
      "text/plain": [
       "[('Germplasm',), ('Locus_Gene',)]"
      ]
     },
     "execution_count": 91,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "%sql show databases;\n",
    "#%sql create database EW2;\n",
    "%sql use EW2;\n",
    "%sql drop table Germplasm\n",
    "%sql drop table Locus_Gene\n",
    "%sql create table Germplasm(Locus VARCHAR(10) NOT NULL, Germplasm VARCHAR(30) NOT NULL, Phenotype VARCHAR(1000) NOT NULL, Pubmed INTEGER NOT NULL);\n",
    "%sql create table Locus_Gene(Germplasm_Locus VARCHAR(10) NOT NULL, Gene VARCHAR(10) NOT NULL, Protein_Length INTEGER NOT NULL);\n",
    "%sql show tables;\n",
    "#%sql describe Germplasm;\n",
    "#%sql describe Locus_Gene;"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Problem 3: Fill the database\n",
    "\n",
    "To do so, I have stored all columns in different lists corresponding to each variable. Then, I've inserted each value to the tables sequentially (all items 1 from each list correspond to the same Germplasm)."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 92,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\n"
     ]
    }
   ],
   "source": [
    "import pymysql.cursors\n",
    "import csv\n",
    "\n",
    "g_loc = []\n",
    "g_germ = []\n",
    "g_phen = []\n",
    "g_pub = []\n",
    "\n",
    "g_tsvfile = open(\"/home/osboxes/BioinformaticsCourseGit/LOM_BioinfoIntro/EXAMS_W2/Germplasm.tsv\",\"r\")\n",
    "g_lines = g_tsvfile.readlines()\n",
    "\n",
    "for line in g_lines:\n",
    "    if \"Locus\" in line:\n",
    "        pass\n",
    "    else:\n",
    "        linestr = line.strip()\n",
    "        linesp = linestr.split(\"\\t\")\n",
    "        g_loc.append(linesp[0])\n",
    "        g_germ.append(linesp[1])\n",
    "        g_phen.append(linesp[2])\n",
    "        g_pub.append(linesp[3])\n",
    "        \n",
    "lg_loc = []\n",
    "lg_gene = []\n",
    "lg_protlen = []\n",
    "\n",
    "lg_tsvfile = open(\"/home/osboxes/BioinformaticsCourseGit/LOM_BioinfoIntro/EXAMS_W2/LocusGene.tsv\")\n",
    "lg_lines = lg_tsvfile.readlines()\n",
    "\n",
    "\n",
    "for line in lg_lines:\n",
    "    if \"Locus\" in line:\n",
    "        pass\n",
    "    else:\n",
    "        linestr = line.strip()\n",
    "        linesp = linestr.split(\"\\t\")\n",
    "        lg_loc.append(linesp[0])\n",
    "        lg_gene.append(linesp[1])\n",
    "        lg_protlen.append(linesp[2])\n",
    "\n",
    "connection = pymysql.connect(host = \"localhost\", user = \"root\" , password = \"root\", db = \"EW2\", charset = \"utf8mb4\", cursorclass = pymysql.cursors.DictCursor)\n",
    "\n",
    "connection.autocommit(True)\n",
    "try:\n",
    "    with connection.cursor() as cursor:\n",
    "        for a in range(len(g_loc)):\n",
    "            sql = \"insert into Germplasm (Locus, Germplasm, Phenotype, Pubmed) values ({},{},{},{})\".format('\"' + g_loc[a] + '\"', '\"' + g_germ[a] + '\"', '\"' + g_phen[a] + '\"',int(g_pub[a]))\n",
    "            cursor.execute(sql)\n",
    "        for b in range(len(lg_loc)):\n",
    "            sql = \"insert into Locus_Gene (Germplasm_Locus, Gene, Protein_Length) values ({},{},{})\".format('\"' + lg_loc[b] + '\"','\"' + lg_gene[b] + '\"',int(lg_protlen[b]))\n",
    "            cursor.execute(sql)\n",
    "finally:\n",
    "    print(\"\")\n",
    "    #connection.close()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Problem 4: Create reports, written to a file"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 153,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Locus\tGermplasm\tPhenotype\tPubmed\tGene\tProtein_Length\n",
      "AT1G01040\tCS3828\tIncreased abundance of miRNA precursors.\t17369351\tDCL1\t332\n",
      "AT1G01060\tlhy-101\tThe mutant plants are hypersensitive to both FRc and Rc light treatments in hypocotyl elongation and exhibits a small reciprocal enlargement in cotyledon area, albeit not statistically significant.\t16891401\tLHY\t290\n",
      "AT1G01140\tSALK_058629\thypersensitive to low potassium media\t17486125\tCIPK9\t223\n",
      "AT1G01220\tSALK_012400C\tfkgp-1 mutants have about 40 times more L-fucose than wild type Arabidopsis plants, but the levels of other monosaccharides do not appear to differ significantly in the mutants. No obvious phenotypic abnormalities were observed in the fkgp-1 mutants, nor were any differences in the sugar composition of cell wall polysaccharides detected.\t18199744\tFKGP\t190\n",
      "AT2G03720\tSALK_042433\tMultiple straight hairs\t16367956\tMRH6\t189\n",
      "AT2G03800\tgek1-1\tEthanol hypersensitivity.\t15215505\tGEK1\t196\n",
      "AT2G04240\txerico\tResistant to exogenous ABA. Seeds contained lower amounts of endogenous ABA than wildtype.\t17933900\tXERICO\t256\n",
      "AT2G05210\tpot1-1\tNo visible phenotype.\t17627276\tPOT1A\t221\n",
      "AT3G02130\trpk2-2\tThe homozygous progeny is indistinguishable from wild-type plants during vegetative growth but showed several morphological alterations after bolting. These plants displayed enhanced inflorescence branching and formed three times as many siliques and flowers as did wild-type plants.\t17419837\tRPK2\t284\n",
      "AT3G02140\tafp4-1\tDecreased germination on high concentrations of glucose and sorbitol.\t18484180\tTMAC2\t300\n",
      "AT3G02230\trgp1-1\trgp1-1 mutants have significantly lower levels of UDP-L-arabinose mutase activity compared to wild-type plants and significantly lower levels of arabinose in their cell walls.\t21478444\tRGP1\t301\n",
      "AT3G02260\ttir3-1 RGLG1:rglg1 rglg2\tThe triple homozygous progeny has low viability, accumulated anthocyanin, and all plants died before shoot emergence.\t17586653\tBIG\t279\n",
      "AT3G02310\tsep2-1\tNon-described subtle phenotype.\t10821278\tSEP2\t175\n",
      "AT3G02680\tatnbs1-1\tSignificantly smaller when grown in the presence of methyl methanosulfonate (MMS) with root growth.  Normal growth under standard growth conditions.\t17672843\tNBS1\t190\n",
      "AT3G02850\tCS3816\tThe skor-1 mutant is sensitive to toxic cations in addition to K+ depletion.\t17568770\tSKOR\t234\n",
      "AT3G02870\tvtc4-1\tascorbate deficient\t16595667\tVTC4\t311\n",
      "AT3G03260\thdg8-1\tNo visible phenotype.\t16778018\tHDG8\t194\n",
      "AT4G14790\tpdd17\tDefective pollen development.\t19237690\tSUV3\t312\n",
      "AT4G15210\tbmy1-2\tPlants cold-shocked for 6h have an increased starch content compared to wildtype.\t16297066\tBAM5\t313\n",
      "AT4G15560\tcla1-1\tMutant seeds grown on medium supplemented with non-phosphorylated synthetic 1-deoxy-D-xylulose (DX) develop green leaves.\t10982425\tDXS\t219\n",
      "AT4G15570\tmaa3\tHomozygotes are not recovered. Female gametophyte development is delayed and asynchronous. During fertilization, fusion of polar nuclei does not occur. Polar nuclei nucloeli are smaller than WT.\t18772186\tMAA3\t294\n",
      "AT4G15802\tAthspb-2\tEarly flowering, reduced fertility, aborted seeds.\t20388662\tHSBP\t254\n",
      "AT4G15880\tesd4-2\tDecreased mRNA levels of the floral repressors FLC and MAF4 and increased mRNA levels of the floral activators FT and SOC1.\t17513499\tESD4\t265\n",
      "AT4G16420\tprz1-1\tAltered response to auxin and cytokinin\t12747832\tADA2B\t279\n",
      "AT4G16480\tatint4-2\tNo visible phenotype.\t16603666\tINT4\t284\n",
      "AT5G10480\tpas2-3\tSegregates 25% embryo lethal.\t18799749\tPAS2\t301\n",
      "AT5G10510\tplt3-1\tShort roots and shortened root meristem.\t17960244\tAIL6\t310\n",
      "AT5G11110\tkns2\tDefects are specific to pollen exine structure. Smaller mesh size in the exine structure. Increased number of baculae. Fully fertile.\t18779216\tSPS2\t232\n",
      "AT5G11260\thy5-101\tUnder FRc conditions, the length mutant hypocotyls is increased compared to that of wild-type plants.  Under Rc conditions, the hypocotyl length is also increased and the cotyledon area is smaller.\t16891401\tHY5\t221\n",
      "AT5G11510\tmyb3r4-1\tNo visible phenotype.\t17287251\tMYB3R-4\t336\n",
      "AT5G12200\tpyd2-2\tThe pyd2-2 mutant has a wild-type appearance under normal growth conditions. Pyrimidine nucleotide and uridine levels are not changed in the mutant, but uracil levels are increased. These mutants cannot grow normally when uracil is provided as a sole nitrogen source.\t19413687\tPYD2\t310\n",
      "AT5G13290\tcrn-1\tIncreased meristem size. Vegetative meristems are are 30% larger than wild type. After bolting inflorescence meristems are enlarged and occasionally fasciated. Flowers occasionally produce extra organs in the first 3.\t12345678\tCRN\t189\n",
      "\"\n",
      "\n",
      "\n",
      "\"\n",
      "AT3G02850\tCS3816\tThe skor-1 mutant is sensitive to toxic cations in addition to K+ depletion.\t17568770\tSKOR\t234\n",
      "AT4G15570\tmaa3\tHomozygotes are not recovered. Female gametophyte development is delayed and asynchronous. During fertilization, fusion of polar nuclei does not occur. Polar nuclei nucloeli are smaller than WT.\t18772186\tMAA3\t294\n",
      "\"\n",
      "\n",
      "\n",
      "\"\n",
      "Chr1:\t4\tAverage_Protein_Length:\t258.75\n",
      "Chr2:\t4\tAverage_Protein_Length:\t215.5\n",
      "Chr3:\t9\tAverage_Protein_Length:\t252.0\n",
      "Chr4:\t8\tAverage_Protein_Length:\t277.5\n",
      "Chr5:\t7\tAverage_Protein_Length:\t271.2857142857143\n",
      "\"\n",
      "\n",
      "\n",
      "\"\n",
      "\n"
     ]
    }
   ],
   "source": [
    "import pymysql.cursors\n",
    "import io\n",
    "import csv\n",
    "\n",
    "def extractsql(sql):\n",
    "    cursor.execute(sql)\n",
    "    results = cursor.fetchall()\n",
    "    for result in results:\n",
    "        ddbbwriter.writerow([result[\"Locus\"],result[\"Germplasm\"],result[\"Phenotype\"],result[\"Pubmed\"],result[\"Gene\"],result[\"Protein_Length\"]])\n",
    "    ddbbwriter.writerow([\"\\n\\n\\n\"])\n",
    "\n",
    "#connection.autocommit(True)\n",
    "with connection.cursor() as cursor:\n",
    "    with open(\"/home/osboxes/BioinformaticsCourseGit/LOM_BioinfoIntro/EXAMS_W2/Reports.tsv\",\"w\") as tsvfile:\n",
    "        ddbbwriter = csv.writer(tsvfile,delimiter = \"\\t\", quotechar = '\"')\n",
    "        ddbbwriter.writerow([\"Locus\",\"Germplasm\",\"Phenotype\",\"Pubmed\",\"Gene\",\"Protein_Length\"])\n",
    "        sql1 = \"select * from Germplasm, Locus_Gene where Germplasm.Locus = Locus_Gene.Germplasm_Locus\"\n",
    "        extractsql(sql1)\n",
    "        sql2 = \"select * from Germplasm, Locus_Gene where Germplasm.Locus = Locus_Gene.Germplasm_Locus and Locus_Gene.Gene = 'SKOR' or Germplasm.Locus = Locus_Gene.Germplasm_Locus and Locus_Gene.Gene = 'MAA3'\"\n",
    "        extractsql(sql2)\n",
    "        sql34 = \"select Germplasm.Locus, Locus_Gene.Protein_Length from Germplasm, Locus_Gene where Germplasm.Locus = Locus_Gene.Germplasm_Locus\"\n",
    "        cursor.execute(sql34)\n",
    "        results34 = cursor.fetchall()\n",
    "        Chr1 = 0\n",
    "        Sum1 = 0\n",
    "        Chr2 = 0\n",
    "        Sum2 = 0\n",
    "        Chr3 = 0\n",
    "        Sum3 = 0\n",
    "        Chr4 = 0\n",
    "        Sum4 = 0\n",
    "        Chr5 = 0\n",
    "        Sum5 = 0\n",
    "        for result in results34:\n",
    "            chrnum = int(result[\"Locus\"][2])\n",
    "            protlen = int(result[\"Protein_Length\"])\n",
    "            if chrnum == 1:\n",
    "                Chr1 += 1\n",
    "                Sum1 += protlen\n",
    "            elif chrnum == 2:\n",
    "                Chr2 += 1\n",
    "                Sum2 += protlen\n",
    "            elif chrnum == 3:\n",
    "                Chr3 += 1\n",
    "                Sum3 += protlen\n",
    "            elif chrnum == 4:\n",
    "                Chr4 += 1\n",
    "                Sum4 += protlen\n",
    "            elif chrnum == 5:\n",
    "                Chr5 += 1\n",
    "                Sum5 += protlen\n",
    "        ddbbwriter.writerow([\"Chr1:\",Chr1,\"Average_Protein_Length:\",Sum1/Chr1])\n",
    "        ddbbwriter.writerow([\"Chr2:\",Chr2,\"Average_Protein_Length:\",Sum2/Chr2])\n",
    "        ddbbwriter.writerow([\"Chr3:\",Chr3,\"Average_Protein_Length:\",Sum3/Chr3])\n",
    "        ddbbwriter.writerow([\"Chr4:\",Chr4,\"Average_Protein_Length:\",Sum4/Chr4])\n",
    "        ddbbwriter.writerow([\"Chr5:\",Chr5,\"Average_Protein_Length:\",Sum5/Chr5])\n",
    "        ddbbwriter.writerow([\"\\n\\n\\n\"])\n",
    "    tsvfile.close\n",
    "\n",
    "    tsvfile = io.open(\"/home/osboxes/BioinformaticsCourseGit/LOM_BioinfoIntro/EXAMS_W2/Reports.tsv\")\n",
    "    print(tsvfile.read())\n",
    "    tsvfile.close"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.7.3"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
